ChaCha20Poly1305 multi-key bounds (#46)
cfrg/draft-irtf-cfrg-aead-limits: draft-irtf-cfrg-aead-limits.md

      - ins: V. T. Hoang
      - ins: S. Tessaro
      - ins: A. Thiruvengadam

>   ChaCha20Poly1305-MU:
>     title: "The Security of ChaCha20-Poly1305 in the Multi-User Setting"
>     target: https://doi.org/10.1145/3460120.3484814
>     date: 2021-11
>     author: 
>       - ins: J. P. Degabriele
>       - ins: J. Govinden
>       - ins: F. Günther
>       - ins: K. Paterson

Reviewer: ```suggestion
  ChaCha20Poly1305-MU: DOI.10.1145/3460120.3484814
```

Then we're pulling from the definitions elsewhere.  The result is the same, but we get the formatting consistent.

## AEAD_CHACHA20_POLY1305

The known single-user analyses for AEAD_CHACHA20_POLY1305 {{ChaCha20Poly1305-SU}},
{{ChaCha20Poly1305-MU}} combine the confidentiality and integrity limits into a
single expression, covered below. For this AEAD, n = 512, k = 256, and t = 128;
the length l is the sum of AAD and plaintext (in blocks of 128 bits),
see {{ChaCha20Poly1305-MU}}.

<!-- I've got to say that this is a pretty unsatisfactory situation. -->

<!--
    In {{ChaCha20Poly1305-SU}}, L is |AAD| + |plaintext| + 1; the + 1 is one
    block length encoding.
    
    From {{ChaCha20Poly1305-MU}} Theorem 4.1 / 3.4:
      AE <= v * 2^25 * (l+1) / 2^t
    where t = 128.
    (NB: The bound component "c * l" (for c = 3*2^24) is upper-bounding

>     2^25 * (l+1) for the worst case l = |AAD|+|m| = 2; cf. Theorem 3.4.)

Reviewer 2: > Instead of assuming the ChaCha block function to be a good PRF (resulting in a PRF advantage term), the multi-user analysis idealizes the permutation (hence that term vanishes).

Gotcha 👍  I must have missed that in the paper.

        - 3rd term (../2^2n):  <= 2^-160, negligible.
        - 4th term (../2^(k+n)):  roughly <= (\sigma^2 + 2o(q+v)) / 2^256
          <= 2^-64, negligible.
        - 5th term (2^(-r/2)):  = 2^-48
    
    The 5th term, ensuring that the adversary is d-repeating ({{GCM-MU2}},
    Theorem 4.2), was improved in {{ChaCha20Poly1305-MU}} Theorem 7.7 to
      2^-(\delta * r)
    for which \delta can be chosen as \delta = 2 for d < 2^9.
    As d < 2^9 does not affect the above simplifications, this only makes the
    5th term negligible (2^-192), and allows to omit it.
-->
Protocols with nonce randomization have a limit of:

~~~

> AEA <= (q+v)*l*B / 2^127

Author: Very roughly, `\tilde{m}` is the parameter determining the overall balls-into-bins probability: it is what controls the term `2^-\tilde{m}` added for nonce randomization, where in the proof of the latter we set `\tilde{m} = \delta * r`, where \delta in turn controls the number of repeating nonces d and the number of allowed queries q.

So indeed the term `2^(-\delta * r)` is what emerges from all this.

But when it comes to the clear and obvious explanation of this result, don't blame the reader... :wink:

Concrete multi-key bounds for AEAD_CHACHA20_POLY1305 are given in Theorem 7.8
in {{ChaCha20Poly1305-MU}}, covering protocols with nonce randomization like
TLS 1.3 {{TLS}} and QUIC {{?RFC9001}}.

For this AEAD, n = 512, k = 256, t = 128, and r = 96; the length l is the sum
of AAD and plaintext (in blocks of 128 bits).

### Authenticated Encryption Security Limit {#mu-ccp-ae}

<!--
    From {{ChaCha20Poly1305-MU}} Theorem 7.8; for nonce randomization (XN transform).
    
    Let:
        - d: the max. number of times any nonce is repeated across users
        - \delta: the nonce-randomizer result's parameter
        - d = r * (\delta + 1) - 1 < 2^9, \delta = 2 be fixed, satisfying Theorem 7.8
        - this limits the number of encryption queries to q <= r * 2^(r-1) <= 2^101
        - o, B <= 2^261 as required for Theorem 7.8
    
    We can simplify the Theorem 7.8 advantage bound as follows:
        - 1st term:  v([constant]* l + 3)/2^t
          Via Theorem 3.4, the more precise term is:  v * (2^25 * (l + 1) + 3) / 2^128
          The 3v/2^t summand is dominated by the rest, so we simplify to
            (v * (l + 1)) / 2^103
        
        - 2nd term:  d(o + q)/2^k

>           For d < 2^9 (as above) and o + q <= 2^145, this is dominated by the 1st term;

Author: These constraints are merely derived from comparing the 1st and 2nd terms, without further constraints.
How about the following explanation?

```suggestion
          For d < 2^9 (as above) and o + q <= 2^145, this is dominated by the 1st term;
          [[ 1st term <= 2nd term as long as v * (l + 1)/2^103 <= d(o + q)/2^256;
          i.e., o + q <= v * (l + 1) * 2^153 / d.
          Even for minimal values v = 1 and l = 1 in 1st term, with d < 2^9,
          this holds as long as o + q <= 2^145. ]]
```

{{mu-gcm-ae}} therefore contains the integrity limits.

## AEAD_CHACHA20_POLY1305

Concrete multi-key bounds for AEAD_CHACHA20_POLY1305 are given in Theorem 7.8
in {{ChaCha20Poly1305-MU}}, covering protocols with nonce randomization like
TLS 1.3 {{TLS}} and QUIC {{?RFC9001}}.

For this AEAD, n = 512, k = 256, t = 128, and r = 96; the length l is the sum
of AAD and plaintext (in blocks of 128 bits).

### Authenticated Encryption Security Limit {#mu-ccp-ae}

<!--
    From {{ChaCha20Poly1305-MU}} Theorem 7.8; for nonce randomization (XN transform).
    
    Let:

>         - d: the max. number of times any nonce is repeated across users

Author: The paper's complexity indeed isn't helping with extracting the simplified bounds we want to present here *sigh

\delta is a formal tuning parameter in the underlying stochastic "balls-into-bins" lemma, which plays a role when determining d: the number of times a nonce is repeated across users. This balls-into-bins can be used to determine d because of how nonces are formed with randomization.
Now: the larger \delta, the smaller the 8th term [correct], but also the larger d.

In the paper's discussion, we chose \delta = 2 simply as this makes the 8th term "small enough" to not dominate. 1/2 would be mimicking the GGM analysis, and lead to the 8th term being 2^-48 ("problematically large"). Other values might be reasonable, too; based on what Chris said, we should discuss this.

        - \delta: the nonce-randomizer result's parameter
        - d = r * (\delta + 1) - 1 < 2^9, \delta = 2 be fixed, satisfying Theorem 7.8
        - this limits the number of encryption queries to q <= r * 2^(r-1) <= 2^101

>         - o, B <= 2^261 as required for Theorem 7.8

Reviewer 2: Hrm, I think this should be 2^260 based on the constraints in 7.8.

>         - o, B <= 2^261 as required for Theorem 7.8

Author: Not sure I'm following. For `o` (paper notation: `p`) and `B` (paper notation: `σ_e`) we have the constraint

`o, B <= (n-k [-1])/6 * 2^(n-k) = 42.5 * 2^256 <= 2^261`

(Not that 2^260 vs. 2^261 make a difference here :smiley:)

>         - o, B <= 2^261 as required for Theorem 7.8

Reviewer 2: OK, so it's early here, but `42.5 * 2^256 <= 2^261` isn't true, right? `2^261 < 42.5 * 2^256 <= 2^262`. (I made a mistake in my derivation earlier, but either way, I agree it doesn't matter much)

>         - o, B <= 2^261 as required for Theorem 7.8

Author: My bad, what I wanted to express is `o, B <= 42.5 * 2^256` is satisfied if `o, B <= 2^261`.

    
    We can simplify the Theorem 7.8 advantage bound as follows:
        - 1st term:  v([constant]* l + 3)/2^t
          Via Theorem 3.4, the more precise term is:  v * (2^25 * (l + 1) + 3) / 2^128
          The 3v/2^t summand is dominated by the rest, so we simplify to
            (v * (l + 1)) / 2^103
        
        - 2nd term:  d(o + q)/2^k

>           For d < 2^9 (as above) and o + q <= 2^145, this is dominated by the 1st term;

Reviewer 2: I wasn't able to derive the same constraint on `o+q` here based on the constraints in 7.8. Can we elaborate here?

            we assume that and hence omit the 2nd term.
        
        - 3rd term:  2o * (n - k)/2^k
          This is dominated by the 2nd term; we hence omit it.
        
        - 4th term:  2v * (n - k + 4t)/2^k
          This is dominated by the 1st term; we hence omit it.
        
        - 5th term:  (B + q)^2/2^(n+1)

>           This is dominated by the 1st term as long as B + q < 2^205;

Reviewer 2: This is true, but don't the constraints allow for a larger bound for B+q, e.g., `B+q < 2^261 + 2^128`?

Author: True, but then the 5th term starts dominating and we'd need to involve it in `AEA` and our limits.
I felt like `B + q < 2^205` being a reasonable narrowed constraint that allows us to omit this 5th term.

Reviewer 2: I'm OK with that outcome.

          i.e., negligible and we hence omit it.
        
        - 6th term:  1/2^(2t-2) = 2^-126
          This is negligible, we hence omit it.
        
        - 7th term:  1/2^(n - k - 2) = 2^-254
          This is negligible, we hence omit it.
        
        - 8th term:  1/(\delta * r)
          This is 2^-192 for the chosen \delta = 2, hence negligible and we omit it.
-->

Protocols with nonce randomization have a limit of:

~~~
AEA <= (v * (l + 1)) / 2^103
~~~

It implies the following limit:

~~~
v <= (p * 2^103) / (l + 1)
~~~

Note that this is the same limit as in the single-user case except that the
total number of forgery attempts `v` and maximum message length in blocks `l`
is calculated across all used keys.

### Confidentiality Limit

<!--
    From {{ChaCha20Poly1305-MU}} Theorem 7.8

>     substracting terms for Pr[Bad_5] and Pr[Bad_6],

Reviewer: ```suggestion
    subtracting terms for Pr[Bad_5] and Pr[Bad_6],
```

    and applying simplifications as above (note there are no verification queries),
    the remaining relevant terms are:
    
        - 2nd term:  d(o + q)/2^k
          As d < 2^9, this is upper bounded by   (o+q)/2^247
        
        - 3rd term:  2o * (n - k)/2^k
          This is  o/2^247 , dominated by the 2nd term; we hence omit it.
        
        - 5th term:  (B + q)^2/2^(n+1)
          This is dominated by the 2nd term as long as B + q < sqrt(o+q) * 2^133;

>           i.e., likely negligible in comparison, but we include it as both are small.

Reviewer: See below.  I think that this can be safely omitted.

        
        - 8th term:  1/(\delta * r)
          This is 2^-192 for the chosen \delta = 2, hence negligible and we omit it.
-->

While the AE advantage is dominated by the number of forgery attempts `v`,
those are irrelevant for the confidentiality advantage. The relevant
limit for protocols with nonce randomization becomes dominated, at a very low
level, by the adversary's offline work `o`, number of protected messages `q`
and maximum blocks encrypted `B`, across all used keys:

~~~
CA <= ((o + q) / 2^247) + ((B + q)^2 / 2^513)
~~~

<!--
    Simplifying
      p >= ((o + q) / 2^247) + ((B + q)^2 / 2^513)

    to

      p/2 >= (o + q) / 2^247
      AND
      p/2 >= (B + q)^2 / 2^513

    yields

      q <= (p * 2^246) - o
      AND
      q <= sqrt(p) * 2^256 - B
    
    In addition, the restrictions on q from {{ChaCha20Poly1305-MU}} Theorem 7.8
    applies: q <= r * 2^(r-1) <= 2^101.
    We round this to 2^100; this value can be slightly increased trading off d.
-->

It implies the following simplified limit, which for most reasonable values
of `p` is dominated by a technical limitation around `q = 2^100`:

~~~

> q <= min( p * 2^246 - o,  sqrt(p) * 2^256 - B, 2^100 )

Reviewer 2: This is a good catch and seems to hold for most reasonable values of `p` -- I think we should take it.

> q <= min( p * 2^246 - o,  sqrt(p) * 2^256 - B, 2^100 )

Author: Strictly speaking, `o` is the adversary's offline work while `B` is work done by normal users executing the protocol. So in principle, `B` could be larger than `o`.

Ultimately, the two first min-terms are most likely dominated by 2^100 except for very small `p` or very large `B` or `o`.

> q <= min( p * 2^246 - o,  sqrt(p) * 2^256 - B, 2^100 )

Reviewer: `B` is the work of a single user, no?  I assume then that `o` exceeds that as any adversary should be willing to work harder than any single user might.

I agree that the 2^100 is likely dominant here.  `p` has to get pretty small before the work involved gets plausible.

> q <= min( p * 2^246 - o,  sqrt(p) * 2^256 - B, 2^100 )

Author: Yes, `B` is encryptions/work under a single key/by a single user. So I agree that it's okay to assume that `B <= o` even if formally in the models you could have something else.  Unless maybe you're not rekeying?

~~~
v * 2^64 + (2l * (v + q))^2 <= (p / u) * 2^128
~~~

## Multi-Key Examples

An example protocol might choose to aim for a multi-key AEA, CA, and IA that is at
most 2<sup>-50</sup>.  If the messages exchanged in the protocol are at most a
common Internet MTU of around 1500 bytes, then a value for l might be set to
2<sup>7</sup>.  {{ex-table-mu}} shows limits for q and v across all keys that
might be chosen under these conditions.

| AEAD                   | Maximum q                | Maximum v              |
|:-----------------------|-------------------------:|-----------------------:|
| AEAD_AES_128_GCM       | 2<sup>69</sup>/B         | 2<sup>69</sup>/B       |
| AEAD_AES_256_GCM       | 2<sup>69</sup>/B         | 2<sup>69</sup>/B       |
| AEAD_CHACHA20_POLY1305 | 2<sup>100</sup>          | 2<sup>46</sup>         |
| AEAD_AES_128_CCM       | 2<sup>30</sup>/sqrt(u)   | 2<sup>30</sup>/sqrt(u) |
| AEAD_AES_128_CCM_8     | 2<sup>30.9</sup>/sqrt(u) | 2<sup>13</sup>/u |
{: #ex-table-mu title="Example multi-key limits"}

The limits for AEAD_AES_128_GCM, AEAD_AES_256_GCM, AEAD_AES_128_CCM, and
AEAD_AES_128_CCM_8 assume equal proportions for q and v. The limits for

> AEAD_AES_128_GCM, AEAD_AES_256_GCM and AEAD_CHACHA20_POLY1305 assume nonce
> randomization being deployed, like in TLS 1.3 {{TLS}} and QUIC {{?RFC9001}}.

Reviewer: ```suggestion
AEAD_AES_128_GCM, AEAD_AES_256_GCM and AEAD_CHACHA20_POLY1305 assume the use
of nonce randomization, like in TLS 1.3 {{TLS}} and QUIC {{?RFC9001}}.
```

The limits for AEAD_AES_128_GCM and AEAD_AES_256_GCM further depend on the

> maximum number B of 128-bit blocks encrypted by any single key. For example,
> limiting each key to encrypt at most 2^20 (about a million) messages (of size
> around 1500 bytes as above), this results in B = 2^27 and hence limits for
> q and v of 2^42 messages that can be protected resp. forgery attempts that can
> be tolerated.

Reviewer: Maybe instead

```suggestion
maximum number B of 128-bit blocks encrypted by any single key. For example,
limiting the number of messages to at most 2^20 (about a million) results in
B = 2^27, which limits both q and v to 2^42 messages.
```

Author: The 1500 bytes message size <= 2^7 blocks entered my calculation as B = 2^20 * 2^7, so I guess that should still be mentioned?

```suggestion
maximum number B of 128-bit blocks encrypted by any single key. For example,
limiting the number of messages (of size <= 2^7 blocks) to at most 2^20 (about a
million) per key results in B = 2^27, which limits both q and v to 2^42 messages.
```

Only the limits for AEAD_AES_128_CCM and AEAD_AES_128_CCM_8 depend on the number
of used keys u, which further reduces them considerably. If v is limited to 1,
q can be increased to 2<sup>31</sup>/sqrt(u) for both CCM AEADs.

# Security Considerations {#sec-considerations}

The different analyses of AEAD functions that this work is based upon generally
assume that the underlying primitives are ideal.  For example, that the
pseudorandom function (PRF) or pseudorandom permutation (PRP) the AEAD builds

> upon is indistinguishable from a truly random function, resp. permutation.

Reviewer: I don't know what "resp. permutation" means here.  Can you expand?

Reviewer 2: I assume it means "respectively permutation," corresponding to the PRP earlier in the sentence.

Reviewer: Perhaps the way to reframe this is to say:

> For example, that a pseudorandom function (PRF) used by the AEAD is indistinguishable from a perfect random function or that a pseudorandom permutation (PRP) is indistinguishable from a perfect random permutation.

Or something like that.

Author: Yes, that's what I meant; rephrase as suggested.
